recipes/openai_textcat.py

```diff
@@ -0,0 +1,535 @@
+        So we need to be robust.
+        """
+        output = {}
+        for line in text.strip().split("\n"):
+            if line and ":" in line:
+                key, value = line.split(":", 1)
+                output[key.strip().lower()] = value.strip()
+        return output
+
+
+@prodigy.recipe(
+    # fmt: off
+    "textcat.openai.suggest",
+    input_path=("Path to jsonl data to annotate", "positional", None, Path),
+    label=("Label for binary classification", "positional", None, str),
+    lang=("Language to initialize spaCy model", "option", "l", str),
+    negative_bias=("Probability of negative examples to be included in the stream", "option", "B", float),
+    model=("GPT-3 model to use for completion", "option", "m", str),
+    batch_size=("Batch size to send to OpenAI API", "option", "b", int),
+    segment=("Split sentences", "flag", "S", bool),
+    examples_path=("Examples file to help define the task", "option", "e", Path),
+    prompt_path=("Path to jinja2 prompt template", "option", "p", Path),
+    max_examples=("Max examples to include in prompt", "option", "n", int),
+    verbose=("Print extra information to terminal", "flag", "v", bool),
+    # fmt: on
+)
+def textcat_openai_suggest(
+    dataset: str,
+    input_path: Path,
+    label: str,
+    lang: str = "en",
+    negative_bias: float = 0.5,
+    model: str = "text-davinci-003",
+    batch_size: int = 10,
+    segment: bool = False,
+    examples_path: Optional[Path] = None,
+    prompt_path: Path = DEFAULT_PROMPT_PATH,
+    max_examples: int = 2,
+    verbose: bool = False,
+):
+    """Get bulk TextCat suggestions from an OpenAI API, using zero-shot or
+    few-shot learning.  The results can then be corrected using the
+    `textcat.manual` recipe.
+
+    Here, we use ChatGPT to suggest if a particular text talks about a recipe or
+    not. You can use the parameter `--chatgpt-bias` (float, 0 to 1 inclusive) to
+    set how much we prefer seeing examples that ChatGPT thinks as recipes.
+    """
+
+    api_key, api_org = _get_api_credentials(model)
+    examples = _read_prompt_examples(examples_path)
+    if label is None:
+        msg.fail("textcat.teach requires a --label", exits=1)
+    nlp = spacy.blank(lang)
+
+    if segment:
+        nlp.add_pipe("sentencizer")
+
+    openai = OpenAISuggester(
+        openai_model=model,
+        label=label,
+        max_examples=max_examples,
+        prompt_template=_load_template(prompt_path),
+        verbose=verbose,
+        segment=segment,
+        openai_api_key=api_key,
+        openai_api_org=api_org,
+        openai_timeout_s=120,
+        openai_n=10,
+    )
+
+    for eg in examples:
+        openai.add_example(eg)
+
+    stream = list(srsly.read_jsonl(input_path))
+    stream = openai(
+        tqdm.tqdm(stream), batch_size=batch_size, nlp=nlp, negative_bias=negative_bias
+    )
+
+    return {
+        "dataset": dataset,
+        "view_id": "blocks",
+        "stream": stream,
```

Comment:
When I ran this recipe: 

```
python -m prodigy textcat.openai.suggest openai-text-demo doubt-justtext.jsonl excitement  -F recipes/openai_textcat.py
```

I got a labelling interface that does not show the label. 

![CleanShot 2023-01-17 at 12 11 00](https://user-images.githubusercontent.com/1019791/212884428-e682c09a-2124-4164-9983-2262d487e7b9.png)

The recipe uses a [classification interface](https://prodi.gy/docs/api-interfaces#classification) in the blocks, so this should be fixed by adding a `label` key to each item in the stream with the label name in it.

Reply:
I may be missing some context because I wasn't around at the beginning of this recipe, but shouldn't we add an extra HTML block that makes it clear what OpenAI thought of the text. Having the reason around certainly helps, but I wonder if we can visualise the answer from OpenAI more explicitly. 

Maybe something more like below (made this in pixelmator real quick): 

![CleanShot 2023-01-17 at 12 20 11](https://user-images.githubusercontent.com/1019791/212886153-d658b672-572f-466e-a2b5-d94ee779e028.png)

Reply:
I noticed another issue that popped up with the missing label. Imagine that I'd be interested in detecting two emotions. 

```bash
# Excitement
python -m prodigy textcat.openai.suggest openai-text-demo doubt.jsonl excitement  -F recipes/openai_textcat.py

# Anger
python -m prodigy textcat.openai.suggest openai-text-demo doubt.jsonl excitement  -F recipes/openai_textcat.py
```

Then both of these calls should start at the beginning of my `doubt.jsonl` and not skip any examples as OpenAI tries to varify the emotion. However, if there is no label key on the example then the `_task_hash` on the example will be the same for both emotions.

Reply:
^For the current iteration for the PR I removed the `suggest` recipe. I'm not too attached on it and I'm convinced that the `fetch` might be better (although I'm still having retry and connection problems with it at longer batches). 

Given that (assuming we only have `textcat.openai.correct` and `fetch`), then I think modeling this problem should be something like:

```sh
python3 -m prodigy textcat.openai.correct demo-doubt \
    doubt.jsonl \
   templates/textcat_prompt.jinja2 \
   --labels excitement,anger  \
   --exclusive-classes
```